docs/posts/better-pypi-stats/index.qmd

```diff
@@ -0,0 +1,358 @@
+---
+title: "Better PyPI stats with Python"
+author: "Cody Peterson"
+date: "2024-09-03"
+image: thumbnail.png
+categories:
+    - clickhouse
+    - shiny
+---
+
+***Ibis + ClickHouse + Shiny for Python = better PyPI stats.***
+
+## Overview
+
+[PyPI Stats](https://pypistats.org/about) is a great resource for Python package
+download statistics from PyPI. However, it only contains 180 days of data and
+lacks more detailed analysis we might be interested in. In this post, we'll
+build a dynamic Python application for better PyPI stats using
+[ClickHouse](https://github.com/clickhouse/clickhouse) as our data platform,
+[Ibis](https://github.com/ibis-project/ibis) as our Python data interface, and
+[Shiny for Python](https://github.com/posit-dev/py-shiny) as our dashboarding
+tool.
+
+::: {.callout-note title="What about ClickPy?"}
+[ClickPy](https://github.com/ClickHouse/clickpy) is an existing open source and
+reproducible project built on the same data with ClickHouse. The primary
+difference is that ClickPy uses SQL and JavaScript whereas this project is in
+Python. We also focus on different visualizations and metrics.
+:::
+
+## Prerequisites
+
+Install the required dependencies:
+
+```bash
+pip install 'ibis-framework[clickhouse]' plotly
+```
+
+Then run imports and setup:
+
+```{python}
+import ibis
+import plotly.express as px
+import clickhouse_connect
+
+px.defaults.template = "plotly_dark"
+ibis.options.interactive = True
+```
+
+## Connecting to ClickHouse
+
+You can connect to the public ClickHouse playground's PyPI database:
+
+```{python}
+host = "clickpy-clickhouse.clickhouse.com"
+port = 443
+user = "play"
+database = "pypi"
+
+con = ibis.clickhouse.connect(
+    host=host,
+    port=port,
+    user=user,
+    database=database,
+)
+con.list_tables()
+```
+
+## Top packages by downloads
+
+Let's start by looking at the most downloaded packages:
+
+```{python}
+overall_t = con.table("pypi_downloads")
+
+top_k = 10_000
+overall_t = (
+    overall_t.order_by(ibis.desc("count"))
+    .limit(top_k)
+    .mutate(rank=1 + ibis.row_number().over(order_by=ibis.desc("count")))
```

Comment:
Is adding one here necessary? It seems like you're just after the ordering, which is there without the addition of one.